### PR TITLE
Protobuf through java layer

### PR DESCRIPTION
**Pull request by mlathara**
Changes to (mostly) use exportconfiguration protobuf instead of query json files through the java layer. Left some support for using query json through spark layer - will deprecate that in the future.

Also, wanted to call out that I'm creating a shaded jar now, and using that in place of the "jar-with-dependencies". It seems spark has a dependency on protobuf 2.5 so it seems we need the shaded jar when protobuf is introduced in the spark layer.

### CODE REVIEW

File: src/main/cpp/src/config/pb_config.cc [hunk 1]
```diff
@@ -215,6 +215,9 @@ void GenomicsDBConfigBase::read_from_PB(const genomicsdb_pb::ExportConfiguration
 void GenomicsDBConfigBase::read_and_initialize_vid_and_callset_mapping_if_available(
     const genomicsdb_pb::ExportConfiguration* export_config) {
   //Callset mapping and vid mapping
+  if (m_vid_mapper.is_initialized()) {
+    return;
+  }
```
Comment (kgururaj) [inline src/main/cpp/src/config/pb_config.cc]: Is there a scenario where vid_mapper can be initialized before this?
Scratch that - I see the update from loader function call below.
I would recommend moving this check to near line 226:
```C++
if(has_vid_mapping_file()) 
  do_whatever();
else
  if(has_vid_mapping())
    do_whatever();
  else
     if(!m_vid_mapper.is_initialized())
      throw exception();
```
This ensures that any vid mapping defined in the query object has higher priority than the loader config

File: src/main/java/org/genomicsdb/reader/GenomicsDBFeatureReader.java [hunk 2]
```diff
@@ -136,61 +126,36 @@ public void close() throws IOException {
...
+        }
+        else {
+            Optional<Coordinates.ContigInterval> contigInterval = Optional.of(
+                    Coordinates.ContigInterval.newBuilder().setContig(chr).setBegin(start).setEnd(end).build());
+            List<String> chromosomeIntervalArraysPaths = 
+                    resolveChromosomeArrayFolderList(contigInterval);
+            return new GenomicsDBFeatureIterator(this.loaderJSONFile, 
+                    this.exportConfiguration, Optional.of(chromosomeIntervalArraysPaths),
+                    this.featureCodecHeader, this.codec, chr, OptionalInt.of(start), OptionalInt.of(end));
+        }
```
Comment (kgururaj) [inline src/main/java/org/genomicsdb/reader/GenomicsDBFeatureReader.java]: Didn't really review this file - I'm assuming it works. This part of the code gives me nightmares.
Let me know if you need me to look at this.

File: src/resources/genomicsdb_export_config.proto [hunk 3]
```diff
@@ -88,4 +88,6 @@ message ExportConfiguration {
   optional uint32 segment_size = 24;
   optional uint32 combined_vcf_records_buffer_size_limit = 25;
   optional bool disable_file_locking_in_tiledb = 27;
+  optional int64 query_block_size = 28;
+  optional int64 query_block_size_margin = 29;
```
[the remark this answers is not in the document]
Comment (mlathara) [reply src/resources/genomicsdb_export_config.proto]: These are used in the Spark layer to decide how big of a block/chunk/"input split" each spark worker queries.

File: example/src/test_genomicsdb_bcf_generator.cc [hunk 4]
```diff
@@ -76,7 +81,37 @@ int main(int argc, char *argv[]) {
...
+      std::cerr << "Error converting JSON to binary string\n";
+      exit(-1);
+    }
+    delete resolver;
+    auto success = export_config.ParseFromString(json_to_binary_output);
+    if(!success) {
+      std::cerr << "Could not parse query JSON file to protobuf\n";
+      exit(-1);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mlathara) [reply example/src/test_genomicsdb_bcf_generator.cc]: Makes sense, will do

File: src/main/cpp/src/config/pb_config.cc (hunk 1 above)
Comment (mlathara) [reply src/main/cpp/src/config/pb_config.cc]: Will do

File: src/main/java/org/genomicsdb/reader/GenomicsDBFeatureReader.java (hunk 2 above)
Comment (mlathara) [reply src/main/java/org/genomicsdb/reader/GenomicsDBFeatureReader.java]: Hah - well genomicsdb tests pass anyway. I should try the gatk tests as well just to be safe...

File: src/main/java/org/genomicsdb/spark/GenomicsDBInput.java [hunk 5]
```diff
@@ -323,9 +382,133 @@ public static String createTmpQueryFile(String queryJson, GenomicsDBPartitionInf
         throw e;
       }
   
-      obj.forEach((k, v) -> {
-        if (!(k.equals("query_column_ranges") || k.equals("array"))) {
-          amended.put(k, v);
+      obj.forEach((key, val) -> {
```
[the remark this answers is not in the document]
Comment (mlathara) [reply src/main/java/org/genomicsdb/spark/GenomicsDBInput.java]: I'm keeping this around to maintain backward compatibility with the "mapping layer" we have at ODA that stores the genomicsdb json info in postgres. For now, that uses json files to query genomicsdb and Hollis et al have many instances lying around for their different projects. Once that json file route goes away, I plan on removing this leaving only the protobuf route to query.

File: src/resources/genomicsdb_export_config.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (kgururaj) [reply src/resources/genomicsdb_export_config.proto]: Would it be possible to create a new Protobuf message called SparkConfig and add that as a member to ExportConfig? That way it's clear what's Spark and what's core to GenomicsDB.

File: src/main/java/org/genomicsdb/spark/GenomicsDBInput.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kgururaj) [reply src/main/java/org/genomicsdb/spark/GenomicsDBInput.java]: Ah ok

File: src/main/java/org/genomicsdb/reader/GenomicsDBFeatureReader.java (hunk 2 above)
Comment (nalinigans) [reply src/main/java/org/genomicsdb/reader/GenomicsDBFeatureReader.java]: Yes please, try the gatk tests just to be sure we are maintaining compatibility.

File: src/main/java/org/genomicsdb/spark/GenomicsDBInput.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (nalinigans) [reply src/main/java/org/genomicsdb/spark/GenomicsDBInput.java]: Can we add some doc around this code to mention it exists to maintain query compatibility with say GenomicsDB from 1.1.0 and earlier or something like that?

**Pull request merged**